### PR TITLE
[PATCH v8] perf: timer: benchmark performance for timer functions

### PR DESCRIPTION
**Pull request by JoyceKong-Arm**
Benchmark timer performance for the following timer functions:
Set function: odp_timer_start()
Cancel function: odp_timer_cancel()
Scan function: odp_timer_scan_inline()

### CODE REVIEW

**Review by TuomasTaipale — COMMENTED** [review]
The commit message does not seem to match what is actually changing. It should explain the linux-gen change, e.g.:
```
linux-gen: timer: add possibilty to profile inline timer performance

<what and why stuff changes ...>
```

File: test/performance/odp_timer_perf.c [hunk 1]
```diff
@@ -17,6 +17,7 @@
```
[the remark this answers is not in the document]
Comment (JoyceKong-Arm) [reply test/performance/odp_timer_perf.c:17]: The original intention is to add a case which can measure the timer expire/scan performance. Then we can optimize and measure the timer algorithm better.
As timer expire/scan function is not exposed to the outside, I can only implement profiling in linux-gen and use odp_schedule to call these functions and control expire numbers.
Maybe we can rename the "scan mode" to a more generic one?

[the remark this answers is not in the document]
Comment (TuomasTaipale) [reply test/performance/odp_timer_perf.c:17]: Wouldn't the existing 0-mode ("measure odp_schedule() overhead when using timers") already suffice for the scan profiling?

[the remark this answers is not in the document]
Comment (JoyceKong-Arm) [reply test/performance/odp_timer_perf.c:17]: Think existing 0-mode would do many other operations and it may not hit timer_pool_scan. Compare the below results for scan and schedule, we can find that odp_schedule() is much heavier that scan. Existing sched results can not measure the timer algorithm directly.

Maybe we can integrate scan mode and sched mode to make scan results as supplement to sched mode.
What's your opinion about this?

Scan results:
Inline timer profiling for thread 1:
  scan rounds:     2209
  ave scan nsec:   45.1
Inline timer profiling for thread 2:
  scan rounds:     2208
  ave scan nsec:   42.5

Schedule results:
TOTAL (2 workers)
  events:             10
  ave time:           1.10 sec
  ave rounds per sec: 23.98M
  num before:         0
  ave before:         0.0 nsec
  max before:         0 nsec
  num after:          10
  ave after:          3666524.0 nsec
  max after:          3922400 nsec

[the remark this answers is not in the document]
Comment (TuomasTaipale) [reply test/performance/odp_timer_perf.c:17]: Is the "Scan result" section above now with the new 2-mode?

[the remark this answers is not in the document]
Comment (JoyceKong-Arm) [reply test/performance/odp_timer_perf.c:17]: Nop, I enabled mode-0 to show the above "Scan Results" for comparison.

[the remark this answers is not in the document]
Comment (TuomasTaipale) [reply test/performance/odp_timer_perf.c:17]: OK. With the `odp_timer.c` changes one gets the impression that profiling and potential performance improvements are wanted for the `timer_pool_scan_inline()` function. In my local env, I get practically identical "ave scan nsec" regardless of the mode (0-mode or 2-mode), the "scan rounds" differ a lot with default values but that can be controlled with the period and resolution parameters.

Do "ave scan nsec" values vary a lot in your env depending on the mode?

[the remark this answers is not in the document]
Comment (JoyceKong-Arm) [reply test/performance/odp_timer_perf.c:17]: Just like yours, 'ave scan nsec' is almost the same for 0-mode and 2-mode in my env.
Think this is resonable. For both two modes, default timer numbers are the same and ave scan efficiency is positive correlated to timer numbers.

[the remark this answers is not in the document]
Comment (TuomasTaipale) [reply test/performance/odp_timer_perf.c:17]: OK, so no need for the new mode then?

[the remark this answers is not in the document]
Comment (JoyceKong-Arm) [reply test/performance/odp_timer_perf.c:17]: Agree with you, I would integrate the modes in next version.
Many thanks.

File: platform/linux-generic/odp_timer.c [hunk 2]
```diff
@@ -2140,5 +2157,18 @@ int _odp_timer_init_local(void)
...
 int _odp_timer_term_local(void)
 {
+#if CONFIG_TIMER_PROFILE_INLINE
+	if (timer_local.prof_rounds) {
+		int thr = odp_thread_id();
+
+		odp_ticketlock_lock(&timer_global->lock);
+		_ODP_PRINT("Inline timer profiling for thread %i:\n", thr);
+		_ODP_PRINT("  scan rounds:     %" PRIu64 "\n", timer_local.prof_rounds);
+		_ODP_PRINT("  ave scan nsec: %6.1f\n",
```
Comment (TuomasTaipale) [inline platform/linux-generic/odp_timer.c:2167]: I think no need for the explicit field width, these could be just:
```
_ODP_PRINT("  scan rounds:   %" ...
_ODP_PRINT("  ave scan nsec: %.1f\n" ...
```
so that values are output starting at the same column.

File: test/performance/odp_timer_perf.c [hunk 3]
```diff
@@ -129,7 +131,7 @@ static void print_usage(void)
 	       "                           0: Private timer pools\n"
 	       "                           1: Shared timer pools\n"
 	       "  -m, --mode             Select test mode. Default: 0\n"
-	       "                           0: Measure odp_schedule() overhead when using timers\n"
+	       "                           0: Measure odp_schedule() + scan overhead when using timers\n"
```
Comment (TuomasTaipale) [inline test/performance/odp_timer_perf.c:134]: The "+ scan" addition is unnecessary as it is an implementation detail.

File: test/performance/odp_timer_perf.c [hunk 4]
```diff
@@ -45,6 +45,8 @@ typedef struct test_stat_t {
 	uint64_t events;
```
Comment (TuomasTaipale) [inline test/performance/odp_timer_perf.c:45]: The `odp_timer_perf.c` changes should be moved to a separate commit with reasoning why and what is being changed in the respective commit message.

Comment (JoyceKong-Arm) [reply test/performance/odp_timer_perf.c:45]: Thanks for the comment, would address them in next version.

File: test/performance/odp_timer_perf.c [hunk 5]
```diff
@@ -442,9 +440,12 @@ static int set_timers(test_global_t *global)
 			timer_ctx_t *ctx = &global->timer_ctx[i][j];
 			odp_timer_start_t start_param;
 
+			nsec = START_NS + rand() % num_timer * period_ns;
```
Comment (psavol) [inline test/performance/odp_timer_perf.c:443]: using random start times is not related to start/cancel time measurement, so it would be better to split it into separate patch. Usually, it would be better to use a fixed table of random numbers in a test application (instead of rand library call), as then the test pattern would be repeatable independent of OS, library or ODP version.

**Comment by psavol** [comment]
Please split timer test application change into a separate PR, so that the new config option and application changes can be reviewed and merged independently.

**Comment by psavol** [comment]
Since the new config option works with any inline timer usage, commit message does not need to mention timer perf app.

So suggesting to change this:
"...Enable the branch by setting inline and CONFIG_TIMER_PROFILE_INLINE for sched mode in timer perf test"

into some thing like this: "Enable inline timer scan performance testing by setting CONFIG_TIMER_PROFILE_INLINE in odp_config_internal.h."

File: test/performance/odp_timer_perf.c [hunk 6]
```diff
@@ -761,6 +759,29 @@ static int set_cancel_mode_worker(void *arg)
...
 					break;
 				}
+			}
+		}
+		cancel_c2 = odp_cpu_cycles();
+		cancel_diff += odp_cpu_cycles_diff(cancel_c2, cancel_c1);
+
+		set_c1 = odp_cpu_cycles();
+		for (i = 0; i < num_tp; i++) {
+			tp = global->timer_pool[i].tp;
```
Comment (psavol) [inline test/performance/odp_timer_perf.c:769]: Splitting timer cancel and start calls into separate loops changes application dynamics quite a lot. Currently, timer cancel and start calls are back-to-back so that there is very little chance that start call would fail (due to current time being too near).

Maybe this cancel-loop + start-loop test case could be a new mode (-m) in the test application.

File: test/performance/odp_timer_perf.c (hunk 5 above)
Comment (JoyceKong-Arm) [reply test/performance/odp_timer_perf.c:443]: I would like use a fixed table of random numbers for set_timers, as the set_timers function is not only used for start/cancel timer measurement, also used for profiling scan. I think there is no need to write another timer set function for sched mode.

File: platform/linux-generic/odp_timer.c [hunk 7]
```diff
@@ -2140,5 +2158,18 @@ int _odp_timer_init_local(void)
...
+#if CONFIG_TIMER_PROFILE_INLINE
+	if (timer_local.prof_rounds) {
+		int thr = odp_thread_id();
+
+		odp_ticketlock_lock(&timer_global->lock);
+		_ODP_PRINT("Inline timer profiling for thread %i:\n", thr);
+		_ODP_PRINT("  scan rounds:   %" PRIu64 "\n", timer_local.prof_rounds);
+		_ODP_PRINT("  ave scan nsec: %.1f\n",
+			   (double)timer_local.prof_nsec / timer_local.prof_rounds);
+		odp_ticketlock_unlock(&timer_global->lock);
```
Comment (psavol) [inline platform/linux-generic/odp_timer.c:2170]: These locks can be removed if printing is done with a single function call (format can be the same, just pack everything into one call). Also, performance measurement mode would impact less other functionality if it would not use the same, global lock.

File: platform/linux-generic/odp_timer.c [hunk 8]
```diff
@@ -872,7 +876,17 @@ void _odp_timer_run_inline(int dec)
 	}
 
 	/* Check the timer pools. */
+#if CONFIG_TIMER_PROFILE_INLINE
```
Comment (psavol) [inline platform/linux-generic/odp_timer.c:879]: Since CONFIG_TIMER_PROFILE_INLINE is build time constant, normal if / else can be used instead of #if / #else.  Optimizer will remove the unused branch, so there won't be extra code by default.

Benefit of using normal if-clause is that all code is always visible to the compiler and some other change later does not break the profile code from building.

Please change all "#if CONFIG_TIMER_PROFILE_INLINE" to if (CONFIG_TIMER_PROFILE_INLINE), and remove #if from timer_local_t struct.

**Comment by psavol** [comment]
Two things I'd do differently, otherwise OK. Please, add my review-by to the next version.

**Comment by JoyceKong-Arm** [comment]
> Two things I'd do differently, otherwise OK. Please, add my review-by to the next version.

**Comment by JoyceKong-Arm** [comment]
> Two things I'd do differently, otherwise OK. Please, add my review-by to the next version.

Sure, will send out the new version soon.

**Review by psavol — APPROVED** [review]
